core/src/main/java/com/linecorp/armeria/client/logging/ContentPreviewingClientBuilder.java

```diff
@@ -0,0 +1,80 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.logging;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.nio.charset.Charset;
+import java.util.function.Function;
+
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.common.logging.ContentPreviewerFactory;
+import com.linecorp.armeria.common.logging.ContentPreviewingDecoratorBuilder;
+
+/**
+ * Builds a new {@link ContentPreviewingClient} or its decorator function.
+ */
+public final class ContentPreviewingClientBuilder extends ContentPreviewingDecoratorBuilder {
+
+    @Override
+    public ContentPreviewingClientBuilder contentPreview(int length) {
+        return (ContentPreviewingClientBuilder) super.contentPreview(length);
+    }
+
+    @Override
+    public ContentPreviewingClientBuilder contentPreview(int length, Charset defaultCharset) {
+        return (ContentPreviewingClientBuilder) super.contentPreview(length, defaultCharset);
+    }
+
+    @Override
+    public ContentPreviewingClientBuilder contentPreviewerFactory(ContentPreviewerFactory factory) {
+        return (ContentPreviewingClientBuilder) super.contentPreviewerFactory(factory);
+    }
+
+    @Override
+    public ContentPreviewingClientBuilder requestContentPreviewerFactory(
+            ContentPreviewerFactory requestContentPreviewerFactory) {
+        return (ContentPreviewingClientBuilder)
+                super.requestContentPreviewerFactory(requestContentPreviewerFactory);
+    }
+
+    @Override
+    public ContentPreviewingClientBuilder responseContentPreviewerFactory(
+            ContentPreviewerFactory responseContentPreviewerFactory) {
+        return (ContentPreviewingClientBuilder)
+                super.responseContentPreviewerFactory(responseContentPreviewerFactory);
+    }
+
+    /**
+     * Returns a newly-created {@link ContentPreviewingClient} based on the properties of this builder.
+     */
+    public ContentPreviewingClient build(HttpClient delegate) {
+        checkState(requestContentPreviewerFactory() != null || responseContentPreviewerFactory() != null,
+                   "requestContentPreviewerFactory or responseContentPreviewerFactory must be set.");
+        return new ContentPreviewingClient(delegate, requestContentPreviewerFactory(),
+                                           responseContentPreviewerFactory());
+    }
+
+    /**
+     * Returns a newly-created decorator that decorates an {@link HttpClient} with a new
+     * {@link ContentPreviewingClient} based on the properties of this builder.
+     */
+    public Function<? super HttpClient, ContentPreviewingClient> newDecorator() {
+        return this::build;
```

Comment:
Do you think we should check state here too? It might give a user earlier feedback

Reply:
Which state do you mean?

Reply:
I mean the `checkState` in `build` - I think usually that will be called much later than setting the decorator so could be confusing, but here it would be right away.

Reply:
Ah, that's a good point. Let me add that. 😄